### ISSUE TITLE
ndk-gdb cannot set breakpoints on 64-bit phones.

### ISSUE THREAD

**Issue by kakyoism** [android/ndk#462]
--------------------------------------------------------------------------------

#### Description

I'm trying to debug an NDK program on Android (SDK 24, NDK r15b) from macOS Sierra.
The test device is an arm64-v8a one.

After running

`ndk-gdb --verbose --force --launch`

I get:

```
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin14.5.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
warning: Could not load shared library symbols for 142 libraries, e.g. /system/lib64/libcutils.so.
Use the "info sharedlibrary" command to see the complete listing.
Do you need "set solib-search-path" or "set sysroot"?
0x0000007f93f84a20 in syscall ()
   from /Applications/path/to/myproject/obj/local/arm64-v8a/system/lib64/libc.so
```
Then I found that no symbols are loaded at this point:

```
(gdb) i locals
No symbol table info available.
```

And setting breakpoints will just give me:

```
(gdb) b main.cpp:249
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (main.cpp:249) pending.
```

Then I loaded my native app's lib, set a breakpoint, and tried to continue:

```
(gdb) symbol-file /Applications/path/to/myproject/obj/local/arm64-v8a/libDemo.so 
Reading symbols from /Applications/path/to/myproject/obj/local/arm64-v8a/libDemo.so...done.
(gdb) b MyClass::MyMethod
Breakpoint 1 at 0x158354: file jni/../../MyClasses/MyClass.cpp, line 45.
(gdb) c
Continuing.
Warning:
Cannot insert breakpoint 1.
Cannot access memory at address 0x158354

Command aborted.
(gdb) 

```

I played around with my libs/arm64-v8a/gdb.setup and used absolute paths in there:

```
set sysroot
set solib-search-path /Applications/path/to/myproject/obj/local/arm64-v8a /Applications/NVPACK/ndk/platforms/android-21/arch-arm64/usr/lib
```
without luck.

how should I make breakpoints work?

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: r15b
* Build sytem: ndk-build
* Host OS: macOS Sierra
* Compiler: ndk-build of r15b
* ABI: arm64-v8a
* STL: gnustl_static
* SDK API level: 21, <uses-sdk android:minSdkVersion="21" />, target=android-21
* Device API level: 23

**Comment by rprichard** [comment]
When running `ndk-gdb --launch` on an app, the program is stopped before the app's native libraries have loaded, so I'm guessing it's expected that you can't immediately set a breakpoint. (I *do* see that I can set a breakpoint after calling `symbol-file`, but then I also see the two errors you've reported.)

I think setting a pending breakpoint and continuing execution is supposed to work. I think the debugger has an internal breakpoint that hits after each solib is loaded.

I'm seeing a problem on OS X where I get a "Waiting For Debugger" popup that never goes away. The problem apparently doesn't reproduce on Linux.

**Comment by rprichard** [comment]
I filed a separate issue for the "Waiting For Debugger" popup problem, https://github.com/android-ndk/ndk/issues/549.

**Comment by rprichard** [comment]
I did a test targeting a Nexus 5X bullhead running M (23), using ARM64, and it worked:
```
$ ~/android-ndk-r15b/ndk-gdb --launch -v
ADB command used: 'adb -s 00b12f4908789077'
ADB version: Android Debug Bridge version 1.0.39 Revision 3db08f2c6889-android Installed as /usr/local/bin/adb
Using project directory: /Users/rprichard/ndk-samples-mk/hello-jni 
Found package name: com.example.hellojni
Selected target activity: 'com.example.hellojni.HelloJni'
Application ABIs: arm64-v8a, armeabi, armeabi-v7a, mips, mips64, x86, x86_64
Device ABIs: arm64-v8a, armeabi-v7a, armeabi
Selecting ABI: arm64-v8a
Detected pretty printer: none
Found application data directory: /data/data/com.example.hellojni
App gdbserver not found at /data/data/com.example.hellojni/lib/gdbserver, uploading.
Copying gdbserver to /data/data/com.example.hellojni/arm64-gdbserver.
Uploaded gdbserver to /data/data/com.example.hellojni/arm64-gdbserver
Launching activity com.example.hellojni/com.example.hellojni.HelloJni...
Sleeping for 0.25 seconds.
Pulling '/system/bin/app_process64' to '/Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/bin/app_process64'
Pulling '/system/bin/linker64' to '/Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/bin/linker64'
Pulling '/system/lib64/libc.so' to '/Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/lib64/libc.so'
Pulling '/system/lib64/libm.so' to '/Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/lib64/libm.so'
Pulling '/system/lib64/libdl.so' to '/Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/lib64/libdl.so'
Starting gdbserver...
Redirecting gdbserver output to /var/folders/r5/xgtm72zn2vs_ft9zrl0bmtc400g__4/T/gdbclient.log
Starting jdb to unblock application.
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin14.5.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
warning: Could not load shared library symbols for 111 libraries, e.g. /system/lib64/libcutils.so.
Use the "info sharedlibrary" command to see the complete listing.
Do you need "set solib-search-path" or "set sysroot"?
0x0000007f93e9af1c in syscall () from /Users/rprichard/ndk-samples-mk/hello-jni/obj/local/arm64-v8a/system/lib64/libc.so
(gdb) b hello-jni.c:62
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (hello-jni.c:62) pending.
(gdb) c
Continuing.
JDB finished unblocking application.
warning: Could not load shared library symbols for /data/app/com.example.hellojni-1/oat/arm64/base.odex.
Do you need "set solib-search-path" or "set sysroot"?

Thread 1 "xample.hellojni" hit Breakpoint 1, Java_com_example_hellojni_HelloJni_stringFromJNI (env=0x7f90b9b180, thiz=0x7fdde11044)
    at jni/hello-jni.c:62
62	    return (*env)->NewStringUTF(env, "Hello from JNI !  Compiled with ABI " ABI ".");
(gdb) p env
$1 = (JNIEnv *) 0x7f90b9b180
(gdb) c
Continuing.
```
I did see some warnings:
 - Could not load shared library symbols for 111 libraries
 - No symbol table is loaded.  Use the "file" command.
 - warning: Could not load shared library symbols for /data/app/com.example.hellojni-1/oat/arm64/base.odex

On the second time I ran ndk-gdb, I hit the "Waiting For Debugger" bug.

**Comment by enh** [comment]
close as "can't reproduce"?

**Comment by rprichard** [comment]
I was never able to reproduce something as bad as the current title suggests.

I was able to reproduce the behavior where:
 - `symbol-file` allows setting a breakpoint when the native shared library isn't loaded
 - continuing the program produced the `Cannot insert breakpoint 1.` / `Cannot access memory at address 0x...` errors

When I instead created a pending breakpoint (or waited until the lib was loaded), everything worked.

AFAICT, the remaining problem is that using `symbol-file` breaks gdb. I'm not sure how much of a problem that is.

I'd be happy with closing the issue, or I could retitle it.

**Comment by kakyoism** [comment]
There is a significant difference between your repro with Nexus 5X and mine: Your test involves JNI, while mine was with a native-activity app without the Java layer. I know from my experience that breakpoints work better with JNI apps.

**Comment by rprichard** [comment]
I also tried the native-activity NDK sample (from [the android-mk branch](https://github.com/googlesamples/android-ndk/tree/android-mk/native-activity)) and didn't see anything different. `ndk-gdb --launch` interrupts the app during the Java framework's "Waiting For Debugger" stage, before the native app library has been loaded, so it's not immediately possible to set breakpoints. 

A pending breakpoint (e.g. `b android_main`) still works:

```
rprichard@cashew:~/ndk-samples-mk/native-activity$ ~/android-sdk-r25/android-ndk-r16/ndk-gdb  --launch
Redirecting gdbserver output to /tmp/gdbclient.log
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
warning: Could not load shared library symbols for 154 libraries, e.g. /system/lib64/libandroid_runtime.so.
Use the "info sharedlibrary" command to see the complete listing.
Do you need "set solib-search-path" or "set sysroot"?
0x00000078476af1ac in syscall () from /usr/local/google/home/rprichard/ndk-samples-mk/native-activity/obj/local/arm64-v8a/system/lib64/libc.so
(gdb) b android_main
Function "android_main" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (android_main) pending.
(gdb) c
Continuing.
[New Thread 3328.3370]
[New Thread 3328.3368]
[Switching to Thread 3328.3370]

Thread 11 "native_activity" hit Breakpoint 1, android_main (state=0x77bb975680) at jni/main.c:235
235	    app_dummy();
(gdb) 
```

As an improvement to ndk-gdb, maybe `--launch` could notice that it's launching a NativeActivity-based app and run the program until the native library is loaded?

**Comment by DanAlbert** [comment]
Not reproducible.
